### PR TITLE
general routines for particle creation

### PR DESCRIPTION
**Pull request by MaxThevenet**
This PR proposes a general function for particle creation.

A number of processes rely on the creation of particles during the simulation if a condition is met. Assuming a vector `is_tagged` (with 1 element per particle) contains `1` if the process occurs for the particle, `0` otherwise, the following processes need
- **ionization**: if `is_tagged = 1` for `ions`, create 1 particle in `electrons` and increase the ionization level of `ion`.
- **splitting**: if `is_tagged = 1` for `species1`, create 8 particles in `species1`, remove the original particle and modify the position and weight of new particles.
- **QED**: if `is_tagged = 1` for `photons`, create 1 particle in `electrons` and 1 particle in `positrons`.
- **QED** if `is_tagged = 1` for `electrons` (or `positrons`), create 1 particle in `photons`.

(that could also include collisions).

As a first step, this PR generalizes the particle creation routine of ionization. A new template base class `elementaryProcess` is created (it is templated on the process type, e.g., ionization). Given a vector `is_tagged`, its method `createParticles` takes one source and multiple product particle containers, and does the following things:
- Allocate memory for particles to be created
- initialize a functor (class copyParticle) that copies one particle (this is similar to `AssignGrid` in AMReX)
- Call method `copyAndTransformParticles` that loops over product particle containers, and copy source particle to each of these product containers. Furthermore, it calls functions `transformSourceParticle` and `transformProductParticle`, which modify the source and product particles. These are template functions, to give flexibility.

This way of doing should be flexible enough for most of what we want to do. There are two levels of customization:
- One can create a class derived from `elementaryProcess` with the proper template parameter, and write specializations for `transformSourceParticle` and `transformProductParticle`. That's what is done for `Ionization`, and it may be enough for all processes.
- If newly-created particles require a more complicated processing, one can additionally override function `copyAndTransformParticles` to enable this change (it is a small function basically containing the loop over particles). The goal is really to keep as much as possible in common for all processes.

Note that this is not yet a finalized version:
- the handling of runtime attributes is not very clean yet (we'll finalize that for the QED module)
- ~~I am not sure this runs on GPU (I get a SEGFAULT, but it may be due to the Redistribute that is incompatible with runtime attributes)~~
- ~~So far, only one particle can be created per product species. This is easy to change (and should be the only change required for particle splitting).~~
- ~~I think the `amrex::ParallelFor` still contains some copies that are not necessary. I have not managed to remove all of them and still be able to compile (in particular the copy functor is copied, and I did not manage to fix it).~~

@atmyers @RemiLehe if you have time I would be happy to have your feedback/discuss this PR with you.

### CODE REVIEW

**Comment by MaxThevenet** [comment]
BTW Currently on GPU running `cuda-gdb` on the automated ionization test gives
```
STEP 84 starts ...

CUDA Exception: Warp Misaligned Address

Thread 1 "main2d.gnu.TPRO" received signal CUDA_EXCEPTION_6, Warp Misaligned Address.
[Switching focus to CUDA kernel 0, grid 190550, block (0,0,0), thread (50,0,0), device 0, sm 0, warp 3, lane 18]
```

**Comment by MaxThevenet** [comment]
The ionization automated tests returned error with message
```
STEP 341 starts ...

CUDA Exception: Warp Out-of-range Address
The exception was triggered at PC 0x1301e670

Thread 1 "main2d.gnu.TPRO" received signal CUDA_EXCEPTION_5, Warp Out-of-range Address.
[Switching focus to CUDA kernel 1, grid 755834, block (2,0,0), thread (192,0,0), device 0, sm 4, warp 4, lane 0]
0x000000001301e6d0 in void amrex::launch_global<void amrex::For<int, elementaryProcess<(elementaryProcessType)0>::createParticles(int, amrex::MFIter const&, std::unique_ptr<WarpXParticleContainer, std::default_delete<WarpXParticleContainer> >&, amrex::Vector<WarpXParticleContainer*, std::allocator<WarpXParticleContainer*> >, amrex::PODVector<int, amrex::ManagedArenaAllocator<int> >&, amrex::PODVector<int, amrex::ManagedArenaAllocator<int> >)::{lambda(int)#1}, void>(int, elementaryProcess<(elementaryProcessType)0>::createParticles(int, amrex::MFIter const&, std::unique_ptr<WarpXParticleContainer, std::default_delete<WarpXParticleContainer> >&, amrex::Vector<WarpXParticleContainer*, std::allocator<WarpXParticleContainer*> >, amrex::PODVector<int, amrex::ManagedArenaAllocator<int> >&, amrex::PODVector<int, amrex::ManagedArenaAllocator<int> >)::{lambda(int)#1}&&, unsigned long)::{lambda()#1}>(elementaryProcess<(elementaryProcessType)0>::createParticles(int, amrex::MFIter const&, std::unique_ptr<WarpXParticleContainer, std::default_delete<WarpXParticleContainer> >&, amrex::Vector<WarpXParticleContainer*, std::allocator<WarpXParticleContainer*> >, amrex::PODVector<int, amrex::ManagedArenaAllocator<int> >&, amrex::PODVector<int, amrex::ManagedArenaAllocator<int> >)::{lambda(int)#1})<<<(8,1,1),(256,1,1)>>> ()
```
Adding a `amex::Gpu::streamSynchronize()` at the end of the particle creation routine fixed the issue.

File: Source/Particles/ParticleCreation/CopyParticle.H [hunk 1]
```diff
@@ -0,0 +1,90 @@
+#ifndef COPYPARTICLE_H_
+#define COPYPARTICLE_H_
+
+#include "WarpXParticleContainer.H"
+
+/**
+ * \brief Functor to copy one particle
+ *
+ * This is meant to be a very small class stored in managed memory,
```
Comment (atmyers) [inline Source/Particles/ParticleCreation/CopyParticle.H]: The CopyParticle functor itself doesn't have to be in managed memory, does it? It is a small struct that stores pointers to managed memory, but the functor itself can be copied into kernels by value.

Comment (MaxThevenet) [reply Source/Particles/ParticleCreation/CopyParticle.H]: Oh that's correct, I'll change that, thanks!

Comment (MaxThevenet) [reply Source/Particles/ParticleCreation/CopyParticle.H]: Done and tested on CPU and GPU.

**Comment by MaxThevenet** [comment]
Thanks for your review @atmyers. Some addition to this PR could include:
- update particle splitting so that it uses these functions
- Functions to initialize the module and compute the mask could be members of `elementaryProcess` class, for consistency
- These functions currently take one input particle and create multiple output particles in multiple particle containers. We could probably modify the functions so that they take multiple input particles. This way, it could be used for a number of modules including collisions.

**Comment by MaxThevenet** [comment]
@lucafedeli88 if you have time, could you have a look at this PR and make sure that these functions can be used for QED modules?

**Comment by lucafedeli88** [comment]
> @lucafedeli88 if you have time, could you have a look at this PR and make sure that these functions can be used for QED modules?

sure

**Comment by lucafedeli88** [comment]
It seems to me that these functions should be perfectly suitable for the QED modules!

Another QED process which should be added to your list is:
-  **QED** if `is_tagged = 1` for `electrons` (or `positrons`), create 1 particle in `photons`

However, this is extremely similar to ionization, so it should be straightforward to implement.

Just one question: why does the ` createParticles ` method in ` template<elementaryProcessType ProcessT> class elementaryProcess ` contain explicit references to the ionization process (e.g. `BL_PROFILE("createIonizedParticles")` )?

**Comment by MaxThevenet** [comment]
Thanks for looking!
Oops you're right, these were mistakes. Fixed them in last commit.

**Pull request merged**